### PR TITLE
Add Debian/Ubuntu packaging scripts

### PR DESCRIPTION
**Pull request by chenxiaolong**
See: #22

### CODE REVIEW

File: dist/build_source_package.sh [hunk 1]
```diff
@@ -109,6 +109,37 @@ build_pkgbuild() {
     cp -v "${temp_dir}"/pkgbuild/* "${output_dir}"/pkgbuild/
 }
 
+# Build deb source package for Debian/Ubuntu
+build_debsrc() {
```
Comment (chrishoage) [inline dist/build_source_package.sh]: this is super pedantic, sorry, but we're creating a dsc which can be used to create a deb.

**Review by chrishoage — APPROVED** [review]
`dh-exec` is also required. I'm not sure if I should have had it included with the other deb tools but I ran into this error while running the script 

```
dpkg-checkbuilddeps: error: Unmet build dependencies: dh-exec
```

However it worked! 

```
› sudo dpkg -i ipmi-fan-control_0.3.0.r11.gitf9ea589-1_amd64.deb
(Reading database ... 126986 files and directories currently installed.)
Preparing to unpack ipmi-fan-control_0.3.0.r11.gitf9ea589-1_amd64.deb ...
Unpacking ipmi-fan-control (0.3.0.r11.gitf9ea589-1) over (0.3.0) ...
Setting up ipmi-fan-control (0.3.0.r11.gitf9ea589-1) ...
```

it even handled the config files correctly during an upgrade.  

```
› cat /etc/ipmi-fan-control.toml.dpkg-dist 
# Definition of a logical fan zone.
[[zones]]
# IPMI session. If unspecified, the `default` session is used, which runs
# ipmitool with no arguments. Sessions are defined below in the `sessions`
```

File: dist/build_source_package.sh (hunk 1 above)
Comment (chenxiaolong) [reply dist/build_source_package.sh]: Good call. Fixed!

**Comment by chenxiaolong** [comment]
> `dh-exec` is also required. I'm not sure if I should have had it included with the other deb tools but I ran into this error while running the script

`dh-exec` is specified under `Build-Depends` in `dist/debian/control`, but I don't think debuild will automatically install build dependencies. Unless I'm missing something, I think those have to be manually installed before building the package.

EDIT: Looks like `mk-build-deps` can parse `debian/control` and install the required build dependencies: https://serverfault.com/questions/127625/given-a-debian-source-package-how-do-i-install-the-build-deps

**Comment by chrishoage** [comment]
Good to know about `mk-build-deps` thanks!

Edit: Thinking on this a little more, `dh-exec` was required simply to run `./build_source_package.sh -t dsc` at all, so it's a bit of a chicken and egg problem to know if deps are missing

**Comment by chenxiaolong** [comment]
> Edit: Thinking on this a little more, `dh-exec` was required simply to run `./build_source_package.sh -t dsc` at all, so it's a bit of a chicken and egg problem to know if deps are missing

Huh, that's interesting. I was able to reproduce that in a fresh `ubuntu:20.04` docker container. It looks like the reason `debuild` requires the build deps even for making the source package is because it runs `./debian/rules clean`, which could theoretically do anything that requires the build deps.

Would it be worth updating `check_tools dch debuild` in `build_source_package.sh` to:
```
check_tools cargo dch debuild dh-exec
```
so that it fails faster?

**Comment by chrishoage** [comment]
> so that it fails faster?

That's what I was thinking if it wasn't something silly that I did to myself. Looks like it wasn't me, it was Ubuntu :slightly_smiling_face:

**Comment by chenxiaolong** [comment]
OK cool, I'll update the `check_tools` line then.

---

Re. uploading to launchpad.net, it's more complicated than I thought:

* The target distro must be listed in `debian/changelog`
* The version number must also be different for every distro version because the packages are all placed in the same directory in the repo (but with distro-specific metadata that picks the correct `.deb`s to use)

I have to build multiple source packages with these in `debian/changelog`:
```
# Local builds
 ipmi-fan-control (0.3.0.r8.git2d545e3-1) UNRELEASED; urgency=medium
# Ubuntu 18.04 PPA
 ipmi-fan-control (0.3.0.r8.git2d545e3-1~ubuntu18.04) bionic; urgency=medium
# Ubuntu 20.04 PPA
 ipmi-fan-control (0.3.0.r8.git2d545e3-1~ubuntu20.04) focal; urgency=medium
# Ubuntu 20.10 PPA
 ipmi-fan-control (0.3.0.r8.git2d545e3-1~ubuntu20.10) groovy; urgency=medium
```

I think I can update the `build_source_package.sh` script to work like:
```
# Local builds
./dist/build_source_package.sh -t dsc
# Uploadable builds
./dist/build_source_package.sh -t dsc -d bionic
./dist/build_source_package.sh -t dsc -d focal
./dist/build_source_package.sh -t dsc -d groovy
```

One thing I'll have to figure out is how to determine the version suffix (`~ubuntu20.10`) from the version name (eg. `groovy`). The launchpad docs explicitly discourage using a version suffix like `~groovy`. I don't want to hardcode a map of version numbers into the script.

---

EDIT: On second thought, maybe I just do:
```
# Local builds
./dist/build_source_package.sh -t dsc
# Uploadable builds
./dist/build_source_package.sh -t dsc -d bionic -s '~ubuntu18.04'
./dist/build_source_package.sh -t dsc -d focal -s '~ubuntu20.04'
./dist/build_source_package.sh -t dsc -d groovy -s '~ubuntu20.10'
```
That would go in the GitHub Actions build pipeline. The script wouldn't need to know about distros versions as all.

**Comment by chrishoage** [comment]
>  it's more complicated than I thought

Sounds like Debian distros in a nutshell :wink: 

> I don't want to hardcode a map of version numbers into the script.

Depending on how you were planning on building these you could either pass in the version (e.g. `-d focal -dv 20.04`?

>  Edit: On second thought

Edit: Looks like we had the same thought!

**Comment by chenxiaolong** [comment]
> Depending on how you were planning on building these you could either pass in the version (e.g. -d focal -dv 20.04?

Ha! I just had the same idea 😄 I'll do that then.

**Comment by chenxiaolong** [comment]
Hit another roadblock :( Looks like launchpad.net will not accept unsigned source packages.

I'm not comfortable with uploading my own GPG key anywhere, so I think I need to:

1. Create a launchpad.net bot account.
2. Create an `ipmi-fan-control` team in launchpad because individual PPAs don't allow granting permissions to other users.
3. Create a new GPG key for the bot account and register the public key in launchpad.net and the private key in this repo's GitHub Actions secrets.
4. Update `build_source_package.sh` to not set `DEBFULLNAME` and `DEBEMAIL` so that the bot account's name and email can be used.
5. Update `build_source_package.sh` to make it possible to run `debuild` without `-us -uc`.

---

I wonder if it's even worth it to use launchpad.net. It feels like a pain to do anything automated with it.

There's also https://build.opensuse.org/, where a single SRPM + dsc + PKGBUILD can be uploaded and it'll build packages and create repos for all supported versions of Fedora/CentOS/Debian/Ubuntu/Arch.

**Comment by chrishoage** [comment]
> I'm not comfortable with uploading my own GPG key anywhere

Will they allow a [subkey ](https://wiki.debian.org/Subkeys) to be uploaded instead? 

> I wonder if it's even worth it to use launchpad.net. It feels like a pain to do anything automated with it.

Hey now you're starting to sound like me :wink: 

Honestly, with something that gets updated so infrequently it seems fine to just have people install directly :woman_shrugging: 

> There's also https://build.opensuse.org/, where a single SRPM + dsc + PKGBUILD can be uploaded and it'll build packages and create repos for all supported versions of Fedora/CentOS/Debian/Ubuntu/Arch.

Perhaps I'm not looking in the right place, but it doesn't seem to create a repo for Debian based distros

>The Build Service will not create a source repository for .deb-based distributions; only the .deb file will be created and published.

https://en.opensuse.org/openSUSE:Build_Service_Debian_builds#Package_repositories

**Comment by chenxiaolong** [comment]
> Hey now you're starting to sound like me 😉

:joy:

> Will they allow a [subkey ](https://wiki.debian.org/Subkeys) to be uploaded instead?

I didn't think they supported that, but the last time I checked was probably 8 years ago. It looks like they added support just a few months after I tried to use it: https://bugs.launchpad.net/launchpad/+bug/1053568

> Perhaps I'm not looking in the right place, but it doesn't seem to create a repo for Debian based distros
> 
> > The Build Service will not create a source repository for .deb-based distributions; only the .deb file will be created and published.
> 
> https://en.opensuse.org/openSUSE:Build_Service_Debian_builds#Package_repositories

I think "source repository" might be referring to `apt-get source`. You'd be able to `apt-get install ipmi-fan-control` normally, but you couldn't use `apt-get source ipmi-fan-control` to download the dsc and build locally (though of course, you could still curl/wget those files from build.opensuse.org).

I'll test it out over the next couple days and see how well it works. If the service is good, then I'll probably use it. We'd get Debian packages for "free" then, as opposed to just Ubuntu on launchpad.

---

EDIT: I don't think launchpad is an option. It doesn't appear to allow network access during the build.

**Comment by chenxiaolong** [comment]
So build.opensuse.org seems to work incredibly well. I set up a test repo to try things out:

* Repo page: https://build.opensuse.org/package/show/home:chenxiaolong:test/ipmi-fan-control
* User friendly landing/setup page: https://software.opensuse.org//download.html?project=home%3Achenxiaolong%3Atest&package=ipmi-fan-control

I was able to successfully build packages for CentOS/Fedora/OpenSUSE/Debian/Ubuntu with a single RPM and (unsigned) dsc.

I had to make some changes to this project to get things to work though:
* build.opensuse.org also does not allow internet access during the build, so I updated `./dist/build_source_package.sh -t tarball` to run `cargo vendor` so that all Rust dependencies' source code is bundled into the tarball.
* The Debian 10 and Ubuntu 20.04 repos only have Rust 1.41, which is too old, so I added some backport patches to make the compilation work.

---

The only thing to figure out is whether it's worth it to add a binary repo for Arch. OBS accepts the PKGBUILD as you have it now (because it references the tarball directly and thus, does not require internet). Maybe the AUR is good enough for Arch--3rd party binary repos seem to be uncommon there.

**Comment by chrishoage** [comment]
> I set up a test repo to try things out

I will setup the apt source this weekend and try it out. 

> Maybe the AUR is good enough for Arch--3rd party binary repos seem to be uncommon there.

While this is true, it's not so unheard of that it doesn't happen. In fact the Arch Wiki has 4 opensuse repos listed in the [unofficial user repositories](https://wiki.archlinux.org/index.php/unofficial_user_repositories) page.

If I were in your position I would use opensuse for everything since it is far simpler to manage everything in one place. 

The nice thing about the AUR is that it's a _user_ repository. Let someone else maintain the PKGBUILD there if they desire :slightly_smiling_face:

**Comment by chenxiaolong** [comment]
Sounds good to me! In that case, I'll keep the PKGBUILD as is and update `build_source_package.sh` to use `sed` if `updpkgsums` is unavailable.

**Comment by chrishoage** [comment]
Tried out the ubuntu test repo today and it appeared to work great!

**Comment by chenxiaolong** [comment]
Awesome! I tested the remaining distros via docker and a remote IPMI session and those seem to work well too. Arch is the only exception, where having `TERM=xterm` set causes ipmitool to print out `\e[?2004l` and `\e[?2004h` sequences (apparently part of "bracketed paste mode" which I've never heard of), which trips up ipmi-fan-control. It's not a problem when starting up via systemd though.

I think this PR should be ready to go. I'll merge this, make a 0.3.1 release (no functional changes) so packages get built, and then update the README to add information about the new packages.

**Pull request merged**